### PR TITLE
Auto-install dependencies for build and deploy and rework for #174

### PR DESCRIPTION
**Pull request by s3bc40**
Related issue: #174 

This is my solution for the issue/feature. It is open to any feedback, to improve in my journey since I have done it with my current experience level. It might not be suited to the original demand, so I am not against closing this PR or improve it if needed!

Thanks for your time!

`just test-all` -> OK
live test -> OK

⚠️ Need to update the doc

---

# Main Features Implemented: Automatic Dependency Management

- Integrated automatic dependency installation when running `build` and `deploy` commands
- System automatically checks and installs missing dependencies
- Handles both PyPI packages and GitHub repositories

# Key Components
## Installation System (mox_install)
- Central function for handling all installations
- Used by build, deploy as requested
- Supports both explicit installation requests and auto-installation

## Dependency Utilities (_dependency_utils.py)
- Smart dependency classification
- Version tracking and management -> `versions.toml` inside each lib folders with Pypi SpecifierSet format
- Handles both PyPI and GitHub package formats

## Package Managers
- PyPI: Uses uv package manager for Python packages -> added preprocessing with dep utilities (inspired from gihub implementation)
- GitHub: Custom implementation for GitHub repository downloads -> reworked but kept Brownie logic.

# Improvements proposed

## Code Organization
- Separated dependency utilities into a dedicated module
- Clear separation between PyPI and GitHub package handling
- Improved code documentation with Sphinx-style docstrings

## Test cases
- unit with install
- integrations
- zksync and live

## Configuration Management
- Version tracking in versions.toml

# Potential next features/refacto

- Auto-delete libs that are not in the `config.dependencies` anymore (removed from `moccasin.toml`)?
- Improve test cases -> I think there is room to add better test case from what I did and also optimize them
- Better code optimization from my PR -> maybe with more experienced Py dev to judge and improve my work
- Possible edge-case where a user can add `[package==0.0.5, package==0.1.0]` when `package==0.1.2` is already installed and triggers the uv pip error (yes the user has to intentionally add this error). But the fact is, maybe add a better filtering at first when we add `conf_dependencies + requirements`?

### CODE REVIEW

**Comment by PatrickAlphaC** [comment]
oooooo exciting!! Give me a few days to review

**Comment by s3bc40** [comment]
Yeah there are a lot of changes to take in. Feel free to ask if anything!

File: moccasin/commands/run.py [hunk 1]
```diff
@@ -7,12 +7,17 @@
...
+from moccasin.config import get_config, get_or_initialize_config
 from moccasin.logging import logger
 
+from .install import mox_install
+
 
 def main(args: Namespace) -> int:
-    initialize_global_config()
+    if not args.no_install:
+        mox_install(args)
```
[the remark this answers is not in the document]
Comment (PatrickAlphaC) [reply moccasin/commands/run.py:18]: Should we also check that `hasattr(args, no_install)`?

Might make sense to make this a utility function? Like `check_no_install`?

File: moccasin/commands/install.py [hunk 2]
```diff
@@ -1,119 +1,204 @@
...
+        _pip_installs(pip_new_or_updated, pip_install_path, quiet)
+    else:
+        logger.info("No pip packages to install")
+
+    # Github installs
+    if len(github_new_or_updated) > 0:
+        logger.info("Checking for new or to update GitHub packages...")
+        _github_installs(github_new_or_updated, github_install_path)
+    else:
+        logger.info("No GitHub packages to install")
```
Comment (PatrickAlphaC) [inline moccasin/commands/install.py:92]: Maybe we add something here... If I have 2 dependencies in my `moccasin.toml`, ie:

```
dependencies = ["snekmate", "patrickalphac/test_repo"]
```

And they are already installed, running `mox build` returns:

```
Running compile command...
Checking for new or to update packages...
Package snekmate is already installed at version 
Using latest version for patrickalphac/test_repo: 0.1.1
patrickalphac/test_repo already installed at version 0.1.1
No pip packages to install
No GitHub packages to install
Compiling 8 contracts to build/...
Skipping contract MockV3Aggregator due to uninitialized.
Done compiling project!
```

This seems like a lot of redundant information:

```
Package snekmate is already installed at version 
Using latest version for patrickalphac/test_repo: 0.1.1
patrickalphac/test_repo already installed at version 0.1.1
No pip packages to install
No GitHub packages to install
```

Maybe it should just be:

```
No pip packages to install
No GitHub packages to install
```

Or we even have a conditional like:

```
Nothing to install
```

Or maybe we just don't say anything <- I think this would be the desired move. So we pass `quiet` to be true when a non-install command installs perhaps.

File: tests/cli/test_cli_compile.py [hunk 3]
```diff
@@ -45,12 +64,16 @@ def test_compile_one(complex_temp_path, complex_cleanup_out_folder, mox_path):
...
+            [mox_path, "build", "BuyMeACoffee.vy", "--no-install"],
             check=True,
             capture_output=True,
             text=True,
         )
     finally:
         os.chdir(current_dir)
+
+    assert not complex_temp_path.joinpath(LIB_GH_PATH).exists()
+    assert not complex_temp_path.joinpath(LIB_PIP_PATH).exists()
```
Comment (PatrickAlphaC) [inline tests/cli/test_cli_compile.py:76]: I'm getting: `RuntimeError: context has already been set` for a lot of tests.

I think there might be a threading issue we are not taking into account somehow. I'm on a macos, what are you on?

**Review by PatrickAlphaC — CHANGES_REQUESTED** [review]
Love the direction, I've left a number of comments on how we can improve this, but this is going in the right direction!

**Comment by PatrickAlphaC** [comment]
> Potential next features/refacto
> Auto-delete libs that are not in the config.dependencies anymore (removed from moccasin.toml)?

Let's not do this

> Improve test cases -> I think there is room to add better test case from what I did and also optimize them

Agreed, but it's ok if we don't optimize them for now. I think we probably could add a test case about multiple installs at once? Maybe we have that and I missed it.

> Better code optimization from my PR -> maybe with more experienced Py dev to judge and improve my work

Yeah, we should aim for this to be pretty performant. For example, we don't want to be making API calls to github if we don't have to, that can waste a LOT of time as a project gets bigger. 

> Possible edge-case where a user can add [package==0.0.5, package==0.1.0] when package==0.1.2 is already installed and triggers the uv pip error (yes the user has to intentionally add this error). But the fact is, maybe add a better filtering at first when we add conf_dependencies + requirements?

yes. If they have strict versioning, we should not update their packages. Or, if they have major version changes we should not auto-update. I left a comment about updates in general.

File: tests/cli/test_cli_compile.py (hunk 3 above)
Comment (s3bc40) [reply tests/cli/test_cli_compile.py:76]: I am on linux (Ubuntu 24) and yes it might explain our difference. I do not have the knowledge though, so I asked ChatGPT just to see if it checks with your assumption: we might have a lead

---
The RuntimeError: context has already been set in pytest typically arises from Python's multiprocessing module when there's an attempt to set the start method more than once. This issue can manifest differently across operating systems due to variations in how they handle process spawning.

### Potential Causes:

1. Multiple Calls to set_start_method: If multiprocessing.set_start_method() is invoked more than once in the codebase, it will raise this error upon subsequent calls.

2. Implicit Setting of Start Method: Calling multiprocessing.get_start_method() before setting the start method can implicitly set it, leading to an error when set_start_method() is called afterward.

### Platform Differences:

The discrepancy between macOS and Linux arises because macOS uses the 'spawn' start method by default, while Linux typically uses 'fork'. The 'spawn' method starts a fresh Python interpreter process, which can lead to the context already being set if not handled properly. Ensuring that the start method is set appropriately and only once can resolve this issue across both platforms.

However, safely forking a multithreaded process is problematic, and it is recommended to use 'fork' carefully in multithreaded environments. The 'fork' method is not available on Windows and macOS, where 'spawn' is the default start method.
---

Maybe we'll have to deal with this by specifying a method, but I won't be able to see if it works. Also, I do not know how `fork` mode will behave on other OS... :fearful: 

I was thinking about Docker containers recently. I do not know if it would be fitted to avoid differences in dev env context... But maybe out scope

**Comment by s3bc40** [comment]
Thanks for the feebacks and I'll respond to each demand :+1: 

I'll make some task on my own

File: moccasin/commands/run.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (s3bc40) [reply moccasin/commands/run.py:18]: Yeah I agree since If we use it in other commands, then centralizing it as a utility would be fitted. 

I'll do it but as a last change compare to the others :+1:

File: moccasin/_dependency_utils.py [hunk 4]
```diff
@@ -25,9 +62,395 @@ def classify_dependency(dependency: str) -> DependencyType:
     return DependencyType.PIP
 
 
-def _write_new_dependencies(
-    new_package_ids: list[str], dependency_type: DependencyType
+def get_new_or_updated_dependencies(
```
[the remark this answers is not in the document]
Comment (s3bc40) [reply moccasin/_dependency_utils.py:65]: Good solution. So if we do this like this:

### `mox compile --no-install`

Should be the easiest to handle. Nothing to check

### `mox compile`

Checking for missing packages means we only compare with `versions.toml` inside each lib folder if the `requirement.name` (pip) or `ORG/REPO` (GH) is matching with the `dependencies`. So just checking the key value from `tomlib` reading to the names of each package. Disregarding update, so no need to request Pypi or Github API.

### `mox compile --update-packages`

Taking into account the version for package. For GH package, there are two cases:
- `ORG/REPO` -> no version, so we always check for the latest and see if it is installed
- `ORG/REPO@VERSION` -> we do not check for latest version, we only check if the package is installed at the right version though

For Pip, maybe it is another story. I'll have to check for SpecifierSet to know if I add it to the update list (+ the one to install i not in version toml).

Depending on the specifier I can add it to the list or not:

| Specifier 	|                      Meaning                     	|
|:---------:	|:------------------------------------------------:	|
| ==        	| Exact version (e.g., requests==2.26.0)           	|
| !=        	| Exclude a specific version (e.g., numpy!=1.21.0) 	|
| >         	| Greater than (e.g., pandas>1.3.0)                	|
| >=        	| Greater than or equal to (e.g., flask>=2.0.1)    	|
| <         	| Less than (e.g., django<4.0)                     	|
| <=        	| Less than or equal to (e.g., tensorflow<=2.9.0)  	|

---

I might modify how I set up `GithubDependency` and `PipDependency` for that. Maybe when I am classifying the packages when merging `config.dep + requirements`, idk. I'll see

File: moccasin/commands/install.py (hunk 2 above)
Comment (s3bc40) [reply moccasin/commands/install.py:92]: I agree by default we could pass `quiet` to `True` by default in other command for install. Installation is the background, except if they ask for `mox install` explicitly.

File: tests/cli/test_cli_compile.py (hunk 3 above)
Comment (PatrickAlphaC) [reply tests/cli/test_cli_compile.py:76]: Well. It should work on mac and/or linux. So this is probably a macos bug, I'll see if I can fix.

File: tests/cli/test_cli_inspect.py [hunk 5]
```diff
@@ -5,17 +5,23 @@
...
-def test_inspect_layout_imports(mox_path, installation_project_config, installation_temp_path):
+def test_inspect_layout_imports(
+    mox_path, installation_project_config, installation_temp_path
+):
     current_dir = Path.cwd()
     try:
         os.chdir(current_dir.joinpath(installation_temp_path))
         result = subprocess.run(
-            [mox_path, "install"], check=True, capture_output=True, text=True
+            [mox_path, "install"], check=True, capture_output=False, text=True
```
Comment (s3bc40) [inline tests/cli/test_cli_inspect.py:15]:
```suggestion
            [mox_path, "install"], check=True, capture_output=True, text=True
```

oops my bad here

File: tests/cli/test_cli_compile.py (hunk 3 above)
Comment (s3bc40) [reply tests/cli/test_cli_compile.py:76]: While testing for the quiet feature, I managed to get into the issue though with my previous tests it was working:
```
Traceback (most recent call last):
  File "/home/s3bc40/gh-projects/moccasin/.venv/bin/mox", line 10, in <module>
    sys.exit(main())
             ^^^^^^
  File "/home/s3bc40/gh-projects/moccasin/moccasin/__init__.py", line 8, in main
    __main__.main(sys.argv[1:])
  File "/home/s3bc40/gh-projects/moccasin/moccasin/__main__.py", line 58, in main
    import_module(f"moccasin.commands.{command_to_run}").main(args)
  File "/home/s3bc40/gh-projects/moccasin/moccasin/commands/compile.py", line 58, in main
    compile_project(
  File "/home/s3bc40/gh-projects/moccasin/moccasin/commands/compile.py", line 128, in compile_project
    multiprocessing.set_start_method(start_method, force=False)
  File "/usr/lib/python3.12/multiprocessing/context.py", line 247, in set_start_method
    raise RuntimeError('context has already been set')
RuntimeError: context has already been set
```

`mox compile` inside the `installation_project` triggers this error ONLY if the `lib` is not created (so if mox install did not run before). I tried it by doing `mox compile` with/without the `lib` folder. Maybe something with `subprocess`?

Comment (s3bc40) [reply tests/cli/test_cli_compile.py:76]: I might have found a way to prevent the issue. It seems like something inside `install.py` is running multiprocessing implicitly. Apparently, with the help of my AI agent, it could be `uv` or `requests` but, uv run inside a subprocess so... maybe not

So to prevent any issue, the suggestion come to my next commit would be to set the `multiprocessing` context as the first process of compile, before install:
```py
def main(args: Namespace) -> int:
    # Set multiprocessing context first
    # @dev check if OS is Windows since fork
    # is not supported on Windows, change it to spawn method
    start_method = "fork" if not IS_WINDOWS else "spawn"
    try:
        multiprocessing.set_start_method(start_method)
    except RuntimeError:
        logger.debug("Multiprocessing context already set by parent process")

    if not args.no_install:
        init_quiet: bool | None = args.quiet if hasattr(args, "quiet") else None
        # Force/add quiet arg for install
        args.quiet = True
        mox_install(args)
        # Remove quiet arg if it was set before install
        if init_quiet is not None:
            args.quiet = init_quiet
        else:
            del args.quiet
```

Doing this prevent the error to comes by. I have let a `logger.debug` to track it, but it might be shown in the output. I can remove it if it is not necessary.

**Comment by s3bc40** [comment]
# feat: quiet output for other commands like compile, run fd1d900a15bf32ada614327b423027545517bcf3 

- output from logging install in other commands are force to quiet while install runs
- helper created to for install process in other commands
- fixed `RuntimeError: context has already been set` when running compile with install by moving multiprocessing set context as first process

just test-all OK
just format done

**Comment by s3bc40** [comment]
## feat: options to command compile and run to optimize installation process https://github.com/Cyfrin/moccasin/pull/204/commits/3235441a46efb12df6e28a7418da2644a7bfd8eb

- install command logs are quiet by default in other commands except `mox install`
- `mox compile --no-install`
- `mox compile` only run install package but no update
- `mox compile --update-packages` will intall and check if updates are available
- install code has been reworked to classify each package in upper level
- conditions has been made to optimize installation process depending on option activated
- used SpecifierSet from python packaging to check for status update pip package

---
just test-all OK
format too

> Note: did not do specific test for run with `--update-packages` but will come later

---
## TODO

- [ ]  add test for run command
- [ ] add install process for `test` and `deploy`
- [ ] update doc
- [ ] push tests with multiple successive install

**Comment by s3bc40** [comment]
## feat: add install to deploy and test and improve tests https://github.com/Cyfrin/moccasin/pull/204/commits/f875d24e0010fcfd558433715ab4bcc983c76182

- ``mox install`` added to ``deploy`` and ``test`` commands
- helper available for rewriting moccasin dependencies while keeping config for tests
- doc cli updated with ``--update-packages`` available
- removed unecessary constant toml tests
- fixed check mox install util

just test-all OK
format done

**Comment by s3bc40** [comment]
## refactor: parsing dep and writing new dependencies https://github.com/Cyfrin/moccasin/pull/204/commits/1d6c3ba590054dbef068ce2addfdb1f2fb9412b3

- reworked parsing
- reworked write dependencies to mocassin toml
- added new tests with ``@pytest.mark.parametrize``

just test-all OK and format

---

Review is possible @PatrickAlphaC  :+1:

**Comment by PatrickAlphaC** [comment]
Thanks! I'll give this a review this week, sorry on my delay.

**Comment by s3bc40** [comment]
No problem you have other priorities 👍

**Comment by PatrickAlphaC** [comment]
going to close this in favor of  https://github.com/Cyfrin/moccasin/pull/214
